### ISSUE TITLE
Is it reasonable that the encoder is so small after FireRedASR/Export_FireRedASR_AED.py?

### ISSUE THREAD

**Issue by InsaneGe** [DakeQQ/Automatic-Speech-Recognition-ASR-ONNX#7]
```bash
(base) root@autodl-container-58da118cfa-c8bc4bb8:~/Automatic-Speech-Recognition-ASR-ONNX-main/FireRedASR# ll -h
-rw-r--r--  1 root root  23K Jul 14 11:44 Export_FireRedASR_AED.py
-rw-r--r--  1 root root 1.5G Jul 14 11:46 FireRedASR_AED_L-Decoder.onnx
-rw-r--r--  1 root root 373K Jul 14 11:45 FireRedASR_AED_L-Encoder.onnx
-rw-r--r--  1 root root  12K Jul 14 10:33 Inference_FireRedASR_AED_ONNX.py
-rw-r--r--  1 root root 7.6K Jul 14 14:36 Optimize_ONNX_q8f32.py
```
FireRedASR_AED_L-Encoder.onnx is only 373K, why could it become so small?
May I ask what is the purpose of "FireRedASR/Optimize_ONNX_q8f32.py"? Is it used to perform additional optimizations on the ONNX model generated by "FireRedASR/Export_FireRedASR_AED.py"? 
can I use it to optimize the [fireredasr onnx](https://huggingface.co/csukuangfj/sherpa-onnx-fire-red-asr-large-zh_en-fp16-2025-02-16/tree/main]). my goal is accelerating the inference process of this ASR model.

**Comment by DakeQQ** [comment]
The 373K file is just an ONNX graph. Due to the optimized code, the model is split into two parts: the graph and the weights. These parts must be used together, or loading will fail.

The script `Optimize_ONNX_q8f32.py` is a tool that includes `onnxslim` and `transformer optimizations`. It improves inference speed by approximately 5%, but the model should work with or without this optimization.

The model link you provided, `sherpa-onnx-fire-red-asr`, appears to already be optimized, so no further optimization is necessary.

**Comment by InsaneGe** [comment]
however, according to the following code, the FireRedASR_AED_L-Encoder.onnx should include the encoder module of the fireredasr model?
```python
        fire_red_encoder = FIRE_RED_ENCODER(model, feat_extractor, custom_stft, NFFT_STFT, N_MELS, SAMPLE_RATE, PRE_EMPHASIZE)
        torch.onnx.export(
            fire_red_encoder,
            (audio,),
            onnx_model_A,
            input_names=['audio'],
            output_names=output_names,
            dynamic_axes=dynamic_axes if DYNAMIC_AXES else None,
            do_constant_folding=True,
            opset_version=17
        )
```

what's more, I have another question I’d like to ask for your advice. In our latency testing, we tested separately on Test Set A and Test Set B, as well as their combined set AB. The results are shown below. I'm wondering why the size of the test set has such a significant impact on the performance?
```bash
==================================================
testA:
==================================================
总请求数：149
总耗时：43.35 s
平均延迟：290.25 ms
P50延迟：135.27 ms
P90延迟：682.70 ms
P95延迟：700.14 ms
P99延迟：1079.65 ms

==================================================
test B:
==================================================
总请求数：148
总耗时：48.06 s
平均延迟：324.04 ms
P50延迟：153.71 ms
P90延迟：692.91 ms
P95延迟：743.22 ms
P99延迟：1416.42 ms

==================================================
test A+B:
==================================================
总请求数：297
总耗时：66.37 s
平均延迟：222.81 ms
P50延迟：119.74 ms
P90延迟：668.50 ms
P95延迟：700.45 ms
P99延迟：887.57 ms
```

Moreover, for audio clips that are nearly the same length (around 1 second), the inference time varies by up to 6 times. Is there any way to further optimize this?
```bash
ASR Result:
"我来啦"
Time Cost: 643.48 ms

ASR Result:
"想要给我"
Time Cost: 127.77 ms
```

Our test script is as follows. I believe the timing measurement is reasonable, as it only measures the model inference time.
```python
import os
import sys
import time
import numpy as np
import onnxruntime
from pydub import AudioSegment


project_path = "/root/FireRedASR"                                    # The FireRedASR Github project path.
model_path = "/root/autodl-tmp/FireRedTeam/FireRedASR-AED-L"                                     # The FireRedASR-AED model download path.
onnx_model_A = "/root/Automatic-Speech-Recognition-ASR-ONNX-main/FireRedASR/FireRedASR_AED_L-Encoder.onnx"      # The exported onnx model path.
onnx_model_B = "/root/Automatic-Speech-Recognition-ASR-ONNX-main/FireRedASR/FireRedASR_AED_L-Decoder.onnx"      # The exported onnx model path.
# test_audio = ["/root/223.wav"]
test_audio = [os.path.join("/root/autodl-tmp/1_processin_A_split", i) for i in os.listdir("/root/autodl-tmp/1_processin_A_split") if i.endswith(".wav")]
test_audio.extend(os.path.join("/root/autodl-tmp/1_sendout_A_split", i) for i in os.listdir("/root/autodl-tmp/1_sendout_A_split") if i.endswith(".wav"))
test_audio[0], test_audio[14] = test_audio[14], test_audio[0]
ORT_Accelerate_Providers = ['CUDAExecutionProvider']           # If you have accelerate devices for : ['CUDAExecutionProvider', 'TensorrtExecutionProvider', 'CoreMLExecutionProvider', 'DmlExecutionProvider', 'OpenVINOExecutionProvider', 'ROCMExecutionProvider', 'MIGraphXExecutionProvider', 'AzureExecutionProvider']
                                        # else keep empty.
MAX_THREADS = 4                         # Max CPU parallel threads.
DEVICE_ID = 0                           # The GPU id, default to 0.
SLIDING_WINDOW = 0                      # Set the sliding window step for test audio reading; use 0 to disable.
MAX_SEQ_LEN = 64                        # It should keep the same with exported model.
SAMPLE_RATE = 16000                     # The model parameter, do not edit the value.
STOP_TOKEN = [4]                        # 4 is the end token for FireRedASR series model.

if "OpenVINOExecutionProvider" in ORT_Accelerate_Providers:
    provider_options = [
        {
            'device_type': 'CPU',                         # [CPU, NPU, GPU, GPU.0, GPU.1]]
            'precision': 'ACCURACY',                      # [FP32, FP16, ACCURACY]
            'num_of_threads': MAX_THREADS,
            'num_streams': 1,
            'enable_opencl_throttling': True,
            'enable_qdq_optimizer': False,                # Enable it carefully
            'disable_dynamic_shapes': False
        }
    ]
    device_type = 'cpu'
elif "CUDAExecutionProvider" in ORT_Accelerate_Providers:
    provider_options = [
        {
            'device_id': DEVICE_ID,
            'gpu_mem_limit': 24 * 1024 * 1024 * 1024,     # 24 GB
            'arena_extend_strategy': 'kNextPowerOfTwo',   # ["kNextPowerOfTwo", "kSameAsRequested"]
            'cudnn_conv_algo_search': 'EXHAUSTIVE',       # ["DEFAULT", "HEURISTIC", "EXHAUSTIVE"]
            'sdpa_kernel': '2',                           # ["0", "1", "2"]
            'use_tf32': '1',
            'fuse_conv_bias': '1',
            'cudnn_conv_use_max_workspace': '1',
            'cudnn_conv1d_pad_to_nc1d': '1',
            'tunable_op_enable': '1',
            'tunable_op_tuning_enable': '1',
            'tunable_op_max_tuning_duration_ms': 10000,
            'do_copy_in_default_stream': '0',
            'enable_cuda_graph': '0',                     # Set to '0' to avoid potential errors when enabled.
            'prefer_nhwc': '0',
            'enable_skip_layer_norm_strict_mode': '0',
            'use_ep_level_unified_stream': '0',
        }
    ]
    device_type = 'cuda'
elif "DmlExecutionProvider" in ORT_Accelerate_Providers:
    provider_options = [
        {
            'device_id': DEVICE_ID,
            'performance_preference': 'high_performance',  # [high_performance, default, minimum_power]
            'device_filter': 'npu'                         # [any, npu, gpu]
        }
    ]
    device_type = 'dml'
else:
    # Please config by yourself for others providers.
    device_type = 'cpu'
    provider_options = None


if project_path not in sys.path:
    sys.path.append(project_path)

from fireredasr.tokenizer.aed_tokenizer import ChineseCharEnglishSpmTokenizer


def normalize_to_int16(audio):
    max_val = np.max(np.abs(audio))
    scaling_factor = 32767.0 / max_val if max_val > 0 else 1.0
    return (audio * float(scaling_factor)).astype(np.int16)


# ONNX Runtime settings
session_opts = onnxruntime.SessionOptions()
session_opts.log_severity_level = 4                   # Fatal level, it an adjustable value.
session_opts.log_verbosity_level = 4                  # Fatal level, it an adjustable value.
session_opts.inter_op_num_threads = MAX_THREADS       # Run different nodes with num_threads. Set 0 for auto.
session_opts.intra_op_num_threads = MAX_THREADS       # Under the node, execute the operators with num_threads. Set 0 for auto.
session_opts.enable_cpu_mem_arena = True              # True for execute speed; False for less memory usage.
session_opts.execution_mode = onnxruntime.ExecutionMode.ORT_SEQUENTIAL
session_opts.graph_optimization_level = onnxruntime.GraphOptimizationLevel.ORT_ENABLE_ALL
session_opts.add_session_config_entry("session.set_denormal_as_zero", "1")
session_opts.add_session_config_entry("session.intra_op.allow_spinning", "1")
session_opts.add_session_config_entry("session.inter_op.allow_spinning", "1")
session_opts.add_session_config_entry("session.enable_quant_qdq_cleanup", "1")
session_opts.add_session_config_entry("session.qdq_matmulnbits_accuracy_level", "4")
session_opts.add_session_config_entry("optimization.enable_gelu_approximation", "1")
session_opts.add_session_config_entry("disable_synchronize_execution_providers", "1")
session_opts.add_session_config_entry("optimization.minimal_build_optimizations", "")
session_opts.add_session_config_entry("session.use_device_allocator_for_initializers", "1")

ort_session_A = onnxruntime.InferenceSession(onnx_model_A, sess_options=session_opts, providers=ORT_Accelerate_Providers, provider_options=provider_options)
print(f"\nUsable Providers: {ort_session_A.get_providers()}")
shape_value_in = ort_session_A._inputs_meta[0].shape[-1]
in_name_A = ort_session_A.get_inputs()
out_name_A = ort_session_A.get_outputs()
in_name_A0 = in_name_A[0].name
output_names_A = []
for i in range(len(out_name_A)):
    output_names_A.append(out_name_A[i].name)

ort_session_B = onnxruntime.InferenceSession(onnx_model_B, sess_options=session_opts, providers=ORT_Accelerate_Providers, provider_options=provider_options)
model_dtype = ort_session_B._inputs_meta[0].type
if 'float16' in model_dtype:
    model_dtype = np.float16
else:
    model_dtype = np.float32
in_name_B = ort_session_B.get_inputs()
out_name_B = ort_session_B.get_outputs()
input_names_B = []
output_names_B = []
amount_of_outputs = len(out_name_B)
for i in range(len(in_name_B)):
    input_names_B.append(in_name_B[i].name)
for i in range(amount_of_outputs):
    output_names_B.append(out_name_B[i].name)

generate_limit = MAX_SEQ_LEN - 1  # 1 = length of input_ids
num_layers = (amount_of_outputs - 2) // 2
num_layers_2 = num_layers + num_layers
num_layers_4 = num_layers_2 + num_layers_2
num_layers_2_plus_1 = num_layers_2 + 1
num_layers_2_plus_2 = num_layers_2 + 2

tokenizer = ChineseCharEnglishSpmTokenizer(model_path + "/dict.txt", model_path + "/train_bpe1000.model")

def calculate_percentile(latencies, percentile):
    return np.percentile(latencies, percentile)
latencies = []
print(f"找到 {len(test_audio)} 个wav文件")
total_start_time = time.time()

# Load the input audio
t=20
for test in test_audio:
    # print("----------------------------------------------------------------------------------------------------------")
    # print(f"\nTest Input Audio: {test}")
    audio = np.array(AudioSegment.from_file(test).set_channels(1).set_frame_rate(SAMPLE_RATE).get_array_of_samples(), dtype=np.float32)
    audio = normalize_to_int16(audio)
    audio_len = len(audio)
    audio = audio.reshape(1, 1, -1)
    if isinstance(shape_value_in, str):
        INPUT_AUDIO_LENGTH = min(320000, audio_len)  # You can adjust it.
    else:
        INPUT_AUDIO_LENGTH = shape_value_in
    if SLIDING_WINDOW <= 0:
        stride_step = INPUT_AUDIO_LENGTH
    else:
        stride_step = SLIDING_WINDOW
    if audio_len > INPUT_AUDIO_LENGTH:
        num_windows = int(np.ceil((audio_len - INPUT_AUDIO_LENGTH) / stride_step)) + 1
        total_length_needed = (num_windows - 1) * stride_step + INPUT_AUDIO_LENGTH
        pad_amount = total_length_needed - audio_len
        final_slice = audio[:, :, -pad_amount:].astype(np.float32)
        white_noise = (np.sqrt(np.mean(final_slice * final_slice)) * np.random.normal(loc=0.0, scale=1.0, size=(1, 1, pad_amount))).astype(audio.dtype)
        audio = np.concatenate((audio, white_noise), axis=-1)
    elif audio_len < INPUT_AUDIO_LENGTH:
        audio_float = audio.astype(np.float32)
        white_noise = (np.sqrt(np.mean(audio_float * audio_float)) * np.random.normal(loc=0.0, scale=1.0, size=(1, 1, INPUT_AUDIO_LENGTH - audio_len))).astype(audio.dtype)
        audio = np.concatenate((audio, white_noise), axis=-1)
    aligned_len = audio.shape[-1]

    # Start to run FireRedASR
    slice_start = 0
    slice_end = INPUT_AUDIO_LENGTH
    input_ids = np.array([[3]], dtype=np.int32)
    ids_len = onnxruntime.OrtValue.ortvalue_from_numpy(np.array([input_ids.shape[-1]], dtype=np.int64), device_type, DEVICE_ID)
    input_ids = onnxruntime.OrtValue.ortvalue_from_numpy(input_ids, device_type, DEVICE_ID)
    history_len = onnxruntime.OrtValue.ortvalue_from_numpy(np.array([0], dtype=np.int64), device_type, DEVICE_ID)
    attention_mask = onnxruntime.OrtValue.ortvalue_from_numpy(np.array([1], dtype=np.int8), device_type, DEVICE_ID)
    past_keys_B = onnxruntime.OrtValue.ortvalue_from_numpy(np.zeros((ort_session_B._inputs_meta[0].shape[0], ort_session_B._inputs_meta[0].shape[1], 0), dtype=model_dtype), device_type, DEVICE_ID)
    past_values_B = onnxruntime.OrtValue.ortvalue_from_numpy(np.zeros((ort_session_B._inputs_meta[num_layers].shape[0], 0, ort_session_B._inputs_meta[num_layers].shape[2]), dtype=model_dtype), device_type, DEVICE_ID)
    layer_indices = np.arange(num_layers_2, num_layers_4, dtype=np.int32) + 3
    input_feed_B = {
        input_names_B[-1]: attention_mask,
        input_names_B[num_layers_2]: input_ids,
        input_names_B[num_layers_2_plus_1]: history_len,
        input_names_B[num_layers_2_plus_2]: ids_len
    }
    for i in range(num_layers):
        input_feed_B[input_names_B[i]] = past_keys_B
    for i in range(num_layers, num_layers_2):
        input_feed_B[input_names_B[i]] = past_values_B
      
    num_decode = 0
    save_token = []
    start_time = time.time()
    while slice_end <= aligned_len:
        all_outputs_A = ort_session_A.run_with_ort_values(output_names_A, {in_name_A0: onnxruntime.OrtValue.ortvalue_from_numpy(audio[:, :, slice_start: slice_end], device_type, DEVICE_ID)})
        for i in range(num_layers_2):
            input_feed_B[input_names_B[layer_indices[i]]] = all_outputs_A[i]
        while num_decode < generate_limit:
            all_outputs_B = ort_session_B.run_with_ort_values(output_names_B, input_feed_B)
            max_logit_ids = onnxruntime.OrtValue.numpy(all_outputs_B[-2])
            num_decode += 1
            if max_logit_ids in STOP_TOKEN:
                break
            for i in range(amount_of_outputs):
                input_feed_B[input_names_B[i]] = all_outputs_B[i]
            if num_decode < 2:
                input_feed_B[input_names_B[-1]] = onnxruntime.OrtValue.ortvalue_from_numpy(np.array([0], dtype=np.int8), device_type, DEVICE_ID)
            save_token.append(max_logit_ids)
        slice_start += stride_step
        slice_end = slice_start + INPUT_AUDIO_LENGTH
    count_time = (time.time() - start_time) * 1000
    text = ("".join([tokenizer.dict[int(id[0][0])] for id in save_token])).replace(tokenizer.SPM_SPACE, ' ').strip()
    latencies.append(count_time)
    if t:
        print(f"\nASR Result:\n{text}\n\nTime Cost: {count_time:.2f} ms\n\nDecode Speed: {num_decode / count_time:.3f} tokens/s")
        t-=1
    # print("----------------------------------------------------------------------------------------------------------")
total_time = time.time() - total_start_time   
# 计算统计信息
total_requests = len(latencies)
avg_latency = np.mean(latencies)
p50_latency = calculate_percentile(latencies, 50)
p90_latency = calculate_percentile(latencies, 90)
p95_latency = calculate_percentile(latencies, 95)
p99_latency = calculate_percentile(latencies, 99)
# 输出统计结果
print("\n" + "="*50)
print("性能统计结果:")
print("="*50)
print(f"总请求数：{total_requests}")
print(f"总耗时：{total_time:.2f} s")
print(f"平均延迟：{avg_latency:.2f} ms")
print(f"P50延迟：{p50_latency:.2f} ms")
print(f"P90延迟：{p90_latency:.2f} ms")
print(f"P95延迟：{p95_latency:.2f} ms")
print(f"P99延迟：{p99_latency:.2f} ms")
if project_path in sys.path:
    sys.path.remove(project_path)
```

**Comment by InsaneGe** [comment]
and why removing the mask in the ConformerEncoder class defined by FireRedASR/modeling_modified/conformer_encoder.py, is it beneficial for exporting the model to ONNX format?
since from our current testing, we’ve observed a drop in accuracy after model compression.
```python
    def forward(self, padded_input, input_lengths=None, pad=True):
        padded_input = torch.cat((padded_input, self.pad_zeros), dim=1)
        enc_output, input_lengths = self.input_preprocessor(padded_input)
        pos_emb = self.positional_encoding(input_lengths)
        enc_outputs = []
        for enc_layer in self.layer_stack:
            enc_output = enc_layer(enc_output, pos_emb, input_lengths, slf_attn_mask=None, pad_mask=None)
            enc_outputs.append(enc_output)
        return enc_output
```
